publishing/s3publisher.py

```diff
@@ -0,0 +1,180 @@
+            site_redirect = SiteRedirect(filename=redirect_filename,
+                                         dir_prefix=directory,
+                                         site_prefix=site_prefix,
+                                         base_url=base_url)
+            local_redirects.append(site_redirect)
+
+    # Combined list of local objects
+    local_objects = local_files + local_redirects
+
+    # Create an S3 client
+    s3_client = boto3.client(
+        service_name='s3',
+        region_name=aws_region,
+        aws_access_key_id=access_key_id,
+        aws_secret_access_key=secret_access_key)
+
+    # Get list of remote files
+    remote_objects = list_remote_objects(bucket=bucket,
+                                         site_prefix=site_prefix,
+                                         s3_client=s3_client)
+
+    # Make dicts by filename of local and remote objects for easier searching
+    remote_objects_by_filename = {}
+    for obj in remote_objects:
+        # These will not have the `directory` prefix that our local
+        # files do, so add it so we can more easily compare them.
+        filename = path.join(directory, obj.filename)
+        if not obj.filename:
+            # Special case where a blank remote filename is the site "root"
+            # redirect object, which we don't want to have a trailing slash.
+            # Instead, it will just have an S3 key name of `directory`.
+            filename = directory
+        remote_objects_by_filename[filename] = obj
+
+    local_objects_by_filename = {}
+    for obj in local_objects:
+        local_objects_by_filename[obj.filename] = obj
+
+    # Create lists of all the new and modified objects
+    new_objects = []
+    modified_objects = []
+    for local_filename, local_obj in local_objects_by_filename.items():
+        matching_remote_obj = remote_objects_by_filename.get(local_filename)
+        if not matching_remote_obj:
+            new_objects.append(local_obj)
+        elif matching_remote_obj.md5 != local_obj.md5:
+            modified_objects.append(local_obj)
+
+    # Create a list of the remote objects that should be deleted
+    deletion_objects = []
+    for remote_filename, remote_obj in remote_objects_by_filename.items():
+        if not local_objects_by_filename.get(remote_filename):
+            deletion_objects.append(remote_obj)
+
+    LOGGER.info('Preparing to upload')
+    LOGGER.info(f'New: {len(new_objects)}')
+    LOGGER.info(f'Modified: {len(modified_objects)}')
+    LOGGER.info(f'Deleted: {len(deletion_objects)}')
+
+    # Upload new and modified files
+    upload_objects = new_objects + modified_objects
+    for file in upload_objects:
+        if dry_run:
+            LOGGER.info(f'Dry-run uploading {file.s3_key}')
+        else:
+            start_time = datetime.now()
+
+            file.upload_to_s3(bucket, s3_client)
+
+            delta = datetime.now() - start_time
+            LOGGER.info(f'Uploaded {file.s3_key} in '
```

Comment:
can use `delta_to_mins_secs` method here

Reply:
I thought about that, but these uploads are often on the order of fractions of a second